flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/FlinkResourceManager.java

```diff
@@ -0,0 +1,801 @@
+
+	// ------------------------------------------------------------------------
+
+	/** The logger, named for the actual implementing class */
+	protected final Logger log = LoggerFactory.getLogger(getClass());
+
+	/** The Flink configuration object */
+	protected final Configuration config;
+
+	/** The timeout for actor messages sent to the JobManager / TaskManagers */
+	private final FiniteDuration messageTimeout;
+
+	/** The service to find the right leader JobManager (to support high availability) */
+	private final LeaderRetrievalService leaderRetriever;
+
+	/** The currently registered resources */
+	private final Map<ResourceID, WorkerType> registeredWorkers;
+
+	/** List of listeners for info messages */
+	private final Set<ActorRef> infoMessageListeners;
+
+	/** The JobManager that the framework master manages resources for */
+	private ActorRef jobManager;
+
+	/** Our JobManager's leader session */
+	private UUID leaderSessionID;
+
+	/** The size of the worker pool that the resource master strives to maintain */
+	private int designatedPoolSize;
+
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Creates a AbstractFrameworkMaster actor.
+	 *
+	 * @param flinkConfig The Flink configuration object.
+	 */
+	protected FlinkResourceManager(
+			int numInitialTaskManagers,
+			Configuration flinkConfig,
+			LeaderRetrievalService leaderRetriever) {
+		this.config = requireNonNull(flinkConfig);
+		this.leaderRetriever = requireNonNull(leaderRetriever);
+		this.registeredWorkers = new HashMap<>();
+
+		FiniteDuration lt;
+		try {
+			lt = AkkaUtils.getLookupTimeout(config);
+		}
+		catch (Exception e) {
+			lt = new FiniteDuration(
+				Duration.apply(ConfigConstants.DEFAULT_AKKA_LOOKUP_TIMEOUT).toMillis(),
+				TimeUnit.MILLISECONDS);
+		}
+		this.messageTimeout = lt;
+		this.designatedPoolSize = numInitialTaskManagers;
+		this.infoMessageListeners = new HashSet<>();
+	}
+
+	// ------------------------------------------------------------------------
+	//  Actor Behavior
+	// ------------------------------------------------------------------------
+
+	@Override
+	public void preStart() {
+		// we start our leader retrieval service to make sure we get informed
+		// about JobManager leader changes
+		try {
+			leaderRetriever.start(new LeaderRetrievalListener() {
+
+				@Override
+				public void notifyLeaderAddress(String leaderAddress, UUID leaderSessionID) {
+					self().tell(
+						new NewLeaderAvailable(leaderAddress, leaderSessionID),
+						ActorRef.noSender());
+				}
+
+				@Override
+				public void handleError(Exception e) {
+					self().tell(
+						new FatalErrorOccurred("Leader retrieval service failed", e),
+						ActorRef.noSender());
+				}
+			});
+		}
+		catch (Throwable t) {
+			self().tell(
+				new FatalErrorOccurred("Could not start leader retrieval service", t),
+				ActorRef.noSender());
+		}
+		// framework specific initialization
+		try {
+			initialize();
```

Comment:
It probably does not make much sense to initialize the resource manager if an error occurred while starting the leader retrieval service.